### ISSUE TITLE
Horizontal flip on COCO

### ISSUE THREAD

**Issue by jnhwkim** [drboog/Lafite#2]
The pre-trained model `COCO2014_CLIP_ViT32_all_text_FID.pkl` has the `'xflip': True`. So, could I assume that to achieve the score in the paper we use horizontal flip but no other augmentation?
Then, `--mirror` should be added in the command line, I assumed.

**Comment by jnhwkim** [comment]
However, I found that, when we use preprocessed clip features, it overrides not to use xflip as in :
https://github.com/drboog/Lafite/blob/main/training/dataset.py#L131

**Comment by drboog** [comment]
1. Yes, the pre-trained model we uploaded used "--mirror=1", and no other data augmentation;
2. I think whether horizontal flip improves the result is dataset-dependent. But to achieve the results in the paper, this is not necessary. We found that we can achieve better results than what we reported in the paper by tuning "--itc" and "--itd", and double training iterations (we haven't update results in the paper). 
3. When you pre-process the dataset, you can also use augmentation before extracting features. I didn't do that because I think horizontal flip will not influence the semantic meaning too much. More importantly,  we will heavily noise the image features to generate fake text features, the impact of augmentation before extracting image feature will be cancelled I think.

**Comment by jnhwkim** [comment]
@drboog What am I asking is the standard setting, which uses text features, not pseudos. Actually, I reimplemented the Lafite myself, and I stuck around FID=15 under the setting of no aug, from-scratch for COCO. Although they are minor differences, I believe the others are minor. In this setting, a horizontal flip might help.

**Comment by drboog** [comment]
Image feature is not used in the standard setting loss, are you using it? Also, may I ask how many kimg have you trained?

**Comment by drboog** [comment]
Also, how many GPUs are you using. According to my experience, itc, itd and gamma matters much more.

**Comment by jnhwkim** [comment]
I am training with the same hyperparams using your code, for COCO, ~17 FID at 3000 kimg. I believe it's okay.

The problem is my own, for my own implementation to reproduce. 
I am reading images from the filesystem which is roughly ~10-15% slow for my environment. 4 GPUs, each has a batch size of 16. I believe it's the same setting. The different things are:
- In your code, for contrastive loss, after gathering, don't backpropagate errors into different ranks ([Ref](https://amsword.medium.com/gradient-backpropagation-with-torch-distributed-all-gather-9f3941a381f8)), but I revised it. I believe this will impact the hyperparameters, since the gradients are 4x (the world size) bigger now in the fixed version, but I'm not sure to use 2.5 instead of 10 for itc and itd.
- Using the exponential moving average of the generator `G_ema` is critical to achieving the FID?

**Comment by drboog** [comment]
Actually, I tried different ways to implement the contrastive loss, including whether to compute the contrastive loss across different GPUs, and whether to use a randomly sampled subset of 64 (4*16) to compute the denominator. 
I quickly tuned the hyper-parameters on a smaller model size, and found that current implementation obtains best results. 
I think other implementations may be able to obtain comparable results, but we have to carefully re-tune the hyper-parameters like temperature (temp), itd, itc, etc.  
I think moving average is important.

**Comment by jnhwkim** [comment]
Yeah, the contrastive loss with `all_gather` is tricky, especially for variable-GPU consistency. I should try `G_ema`.
